### ISSUE TITLE
资源不能被调度

### ISSUE THREAD

**Issue by BeerTai** [OpenRLHF/OpenRLHF#516]
我在利用ppo_ray脚本改为online-dpo时，启动集群时时采用
`ray start --head --num-cpus=4 --num-gpus=4`
在调试一个0.5b模型时，调试脚本如下：
```
ray job submit --address="http://127.0.0.1:8265" \
    --runtime-env-json='{"working_dir": "/OpenRLHF"}' \
    --no-wait \
    -- python3 examples/train_online_dpo_ray.py \
    --ref_num_nodes 1 \
    --ref_num_gpus_per_node 1 \
    --reward_num_nodes 1 \
    --reward_num_gpus_per_node 1 \
    --actor_num_nodes 1 \
    --actor_num_gpus_per_node 1 \
    --vllm_num_engines 1 \
    --vllm_tensor_parallel_size 1 \
```
其中ref和reward共享gpu，为啥一直显示：
```
Warning: The following resource request cannot be scheduled right now: {'GPU': 1.0, 'CPU': 1.0}. This is likely due to all cluster resources being claimed by actors. Consider creating fewer actors or adding more nodes to this Ray cluster.
```
ray status显示如下：
![image](https://github.com/user-attachments/assets/3414a9b3-dab6-4c5f-9732-b3a066913cce)
gpu空闲资源如下：
![image](https://github.com/user-attachments/assets/b7927a0a-4bb1-48c1-85d7-02f30bcd4f53)

**Comment by histmeisah** [comment]
老哥怎么解决的

**Comment by histmeisah** [comment]
> [@BeerTai](https://github.com/BeerTai) [@histmeisah](https://github.com/histmeisah) 同问，怎么解决的呀？我也遇到相同问题了

我改成这样就能跑了，我也没搞懂：
#!/bin/bash

set -x

# 生成训练时间戳
TRAIN_TIMESTAMP=$(date +%Y%m%d_%H%M%S)

# 定义基础路径
BASE_DIR="base_dir"

# 设置环境变量
export MASTER_ADDR=$(hostname -I | awk '{print $1}')
export MASTER_PORT=29500
export WORLD_SIZE=1
export LOCAL_RANK=0
export RANK=0

# 1. 确保Ray完全停止并清理
ray stop
sleep 2
ray stop
sleep 2

# 清理Ray的临时文件
rm -rf /tmp/ray
rm -rf ~/.ray
sleep 2

# 2. 启动Ray (添加更多配置参数)
ray start --head \
    --node-ip-address ${MASTER_ADDR} \
    --port=6379 \
    --dashboard-port=8265 \
    --num-gpus 8 \
    --num-cpus 96 \
    --object-store-memory 50000000000 \
    --redis-max-memory 10000000000

sleep 10  # 增加等待时间

# 验证Ray状态
ray status || exit 1

# 3. 运行训练脚本
cd ${BASE_DIR}

# 定义要排除的文件模式和环境变量
EXCLUDES='{
    "working_dir": "work_path_dir",
    "excludes": [
        "**/*.pt",
        "**/*.bin", 
        "**/*.pth", 
        "**/*.ckpt", 
        "**/*.safetensors",
        "**/wandb/**",
        "**/__pycache__/**"
    ],
    "env_vars": {
        "NCCL_DEBUG": "INFO",
        "NCCL_SOCKET_IFNAME": "enp218s0",
        "GLOO_SOCKET_IFNAME": "enp218s0",
        "NCCL_IB_DISABLE": "1",
        "NCCL_P2P_DISABLE": "1",
        "NCCL_ASYNC_ERROR_HANDLING": "1",
        "NCCL_BLOCKING_WAIT": "1",
        "NCCL_BUFFSIZE": "2097152",
        "CUDA_DEVICE_ORDER": "PCI_BUS_ID",
        "CUDA_VISIBLE_DEVICES": "0,1,2,3,4,5,6,7",
        "CUDA_LAUNCH_BLOCKING": "1",
        "TORCH_DISTRIBUTED_DEBUG": "DETAIL",
        "TORCH_SHOW_CPP_STACKTRACES": "1",
        "TORCH_CUDA_ARCH_LIST": "8.0",
        "RAY_memory_monitor_refresh_ms": "0",
        "TORCH_DISTRIBUTED_BACKEND": "nccl",
        "NCCL_PROTO": "simple"
    }
}'

# 准备训练参数 (优化GPU配置)
TRAIN_ARGS="--ref_num_nodes 1 \
    --ref_num_gpus_per_node 7 \
    --actor_num_nodes 1 \
    --actor_num_gpus_per_node 7 \
    --colocate_actor_ref \
    --vllm_num_engines 1 \
    --vllm_tensor_parallel_size 1 \
    --pretrain pretrain_path \
    --remote_rm_url reward_func.py \
    --save_path model_path \
    --micro_train_batch_size 2 \
    --train_batch_size 28 \
    --micro_rollout_batch_size 2 \
    --rollout_batch_size 84 \
    --max_epochs 2 \
    --prompt_max_len 2048 \
    --generate_max_len 2048 \
    --zero_stage 2 \
    --bf16 \
    --actor_learning_rate 5e-7 \
    --init_kl_coef 0.001 \
    --prompt_data dataset_path \
    --input_key input \
    --advantage_estimator reinforce \
    --adam_offload \
    --apply_chat_template \
    --max_samples 100000 \
    --normalize_reward \
    --gradient_checkpointing \
    --flash_attn \
    --enable_ema \
    --wandb_project qwen2.5-7b-rlhf-test-2025-2-10 \
    --wandb_run_name ${TRAIN_TIMESTAMP}_name \
    --num_episodes 10 \
    --use_wandb [WANDB_TOKENS]"

# 使用ray job submit运行训练
ray job submit --address="http://127.0.0.1:8265" \
    --runtime-env-json="${EXCLUDES}" \
    -- python3 -m openrlhf.cli.train_ppo_ray ${TRAIN_ARGS}

# 4. 等待训练完成并清理
echo "等待训练进程完成..."
sleep 10

# 确保所有进程组被正确销毁
python3 -c "
import torch.distributed as dist
try:
    if dist.is_initialized():
        dist.destroy_process_group()
except:
    pass
"

# 停止Ray
ray stop
sleep 2

# 清理临时文件
rm -rf /tmp/ray
rm -rf ~/.ray

# 清理NCCL临时文件
rm -rf /dev/shm/nccl-*
rm -rf /tmp/nccl-*

# 设置环境变量以避免NCCL警告
export NCCL_ASYNC_ERROR_HANDLING=0
export TORCH_NCCL_ASYNC_ERROR_HANDLING=0

# GPU分配说明:
# 1. 当前配置 (使用colocate_actor_ref, 总共8卡):
#    - Actor和Reference模型共用7张卡 (每张卡75%显存给actor, 25%给ref)
#    - vLLM使用1张卡
#    - 设置了colocate_actor_ref让actor和ref共享GPU以提高利用率
#
# 2. 另一种可选配置 (不使用colocate):
#    --ref_num_nodes 1 \
#    --ref_num_gpus_per_node 1 \
#    --actor_num_nodes 1 \
#    --actor_num_gpus_per_node 6 \
#    (删除--colocate_actor_ref)
#    这种配置下:
#    - Reference模型独占1张卡
#    - Actor模型使用6张卡
#    - vLLM使用1张卡
#
# 3. GPU使用计算公式:
#    - 不使用colocate时: 
#      总GPU = ref_num_nodes*ref_num_gpus_per_node + reward_num_nodes*reward_num_gpus_per_node + critic_num_nodes*critic_num_gpus_per_node + actor_num_nodes*actor_num_gpus_per_node + vllm_num_engines*zero_stage
#    - 使用colocate_critic_reward和colocate_actor_ref时:
#      总GPU = ref_num_nodes*ref_num_gpus_per_node + critic_num_nodes*critic_num_gpus_per_node + vllm_num_engines*zero_stage
#
# 4. 注意事项:
#    - 使用colocate_actor_ref时，ref_num_gpus_per_node必须等于actor_num_gpus_per_node
#    - colocate模式会增加单卡显存使用，但能提高整体GPU利用率
#    - vLLM建议使用单卡以避免张量并行带来的问题

**Comment by BeerTai** [comment]
> [@BeerTai](https://github.com/BeerTai) [@histmeisah](https://github.com/histmeisah) 同问，怎么解决的呀？我也遇到相同问题了

在ray start之前先指定空闲gpu，export CUDA_VISIBLE_DEVICES=0,1

**Comment by linhaojia13** [comment]
@histmeisah @BeerTai 谢谢！已经解决！
